python/hidet/testing/models/llama.py

```diff
@@ -0,0 +1,668 @@
+                "position_ids": position_ids,
+                "past_key_values": past_key_values,
+                "use_cache": kwargs.get("use_cache"),
+                "attention_mask": attention_mask,
+            }
+        )
+        return model_inputs
+
+    @staticmethod
+    def _reorder_cache(past_key_values, beam_idx):
+        reordered_past = ()
+        for layer_past in past_key_values:
+            reordered_past += (tuple(past_state.index_select(0, beam_idx) for past_state in layer_past),)
+        return reordered_past
+
+
+def get_model(args, hf_config):
+    from transformers import AutoModelForCausalLM  # type: ignore[import]
+
+    model_name = args.model
+    model_path = args.model_path
+    dtype = args.quantization.model_dtype
+    max_seq_len = args.max_seq_len
+
+    if model_name.startswith("vicuna-") or model_name.startswith("llama-"):
+        config = LlamaConfig(**hf_config, dtype=dtype)
+        if max_seq_len != -1:
+            config.max_sequence_length = max_seq_len
+
+        hf_model = AutoModelForCausalLM.from_pretrained(model_path)
+        # Get a list of parameters in advance, then delete the model to save memory
+        param_list = [param for _, param in hf_model.named_parameters()]
+        del hf_model
+
+        for i, param in enumerate(param_list):
+            param_list[i] = hidet.from_torch(param.detach()).to(config.dtype)
+
+        # construct the new model without duplicating the weights
+        orig_set_attr = nn.Module.__setattr__
+
+        def my_setattr(self, key, value):
+            # the order of the defined weights are the same
+            if isinstance(value, hidet.Tensor):
+                value = param_list.pop(0)
+            orig_set_attr(self, key, value)
+
+        hidet.nn.Module.__setattr__ = my_setattr
+        hidet_model = LlamaForCausalLM(config)
+        hidet.nn.Module.__setattr__ = orig_set_attr
+
+        return hidet_model
+
+    raise ValueError(f"Unsupported model: {model_name}")
+
+
+def small_model_loading_test():
+    from transformers import LlamaForCausalLM as hfLlamaModel, LlamaConfig as hfLlamaConfig
+
+    configuration = hfLlamaConfig(
+        vocab_size=512, hidden_size=512, intermediate_size=512, num_attention_heads=8, num_hidden_layers=2
+    )
+    hf_model = hfLlamaModel(configuration)
+
+    ###### use the local LlamaConfig
+    config = LlamaConfig(
+        vocab_size=512, hidden_size=512, intermediate_size=512, num_attention_heads=8, num_hidden_layers=2
+    )
+    # Get a list of parameters in advance, then delete the model to save memory
+    param_list = [param for _, param in hf_model.named_parameters()]
+    del hf_model
+
+    for i, param in enumerate(param_list):
+        param_list[i] = hidet.from_torch(param.detach()).to(config.dtype)
+
+    # construct the new model without duplicating the weights
+    # since the hidet model is isomorphic to the original torch model
+    # this only works if we add .register_buffer(...)
```

Comment:
I saw you added `register_buffer`, do we still need this?

Reply:
If we leave it out, the registered buffers will be under the same namespace as the regular parameters in the hidet model, but not the torch one, so we would need to implement model.named_parameters() like torch, to reuse the parameters, which would be more complex. Is there another way otherwise?

Reply:
I have implemented the `named_parameters(...)` in the latest commit: https://github.com/hidet-org/hidet/blob/main/python/hidet/graph/nn/module.py#L96-L103

Does it work for this usage case?

Reply:
For example, 
https://github.com/hidet-org/hidet/blob/6d4bd3d531ba5a839269e3f648baec8179d04f59/python/hidet/testing/models/gpt2.py#L207-L242

constructs the model twice in memory.

Reply:
Good catch!

We can first load the hf model in cpu.